### ISSUE TITLE
StackOverflow in IncrementallyParseTreeAsync

### ISSUE THREAD

**Issue by KirillOsenkov** [dotnet/roslyn#19001]
From @kumpera:

https://bugzilla.xamarin.com/show_bug.cgi?id=55597

1) checkout mono
2) open mcs/class/corlib/corlib-net_4_x.csproj
3) open string.cs
4) scroll to the bottom and wait for Roslyn analysis to show up - tons of yellow stuff will light up on the right side
5) start typing on the last line

After some typing (20ish words) I get a stack overflow followed by a hard crash on this stack:

```
(lldb) bt 40
* thread #114, name = 'Threadpool worker', stop reason = EXC_BAD_ACCESS (code=2, address=0x7000042c0ff8)
  * frame #0: Mono`System.Environment:get_CurrentManagedThreadId () at Environment.cs:163
    frame #1: 0x0000000147e77851 Mono`Roslyn.Utilities.NonReentrantLock:get_IsOwnedByMe () + 33
    frame #2: 0x0000000147e773bf Mono`Roslyn.Utilities.NonReentrantLock:Wait (System.Threading.CancellationToken) + 111
    frame #3: 0x000000014801bdeb Mono`Roslyn.Utilities.AsyncLazy`1<T_REF>:TakeLock (System.Threading.CancellationToken) + 107
    frame #4: 0x000000014801b6e8 Mono`Roslyn.Utilities.AsyncLazy`1<T_REF>:GetValueAsync (System.Threading.CancellationToken) + 728
    frame #5: 0x000000014b7a189c Mono`Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13:MoveNext () + 1020
    frame #6: Mono`System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<Microsoft.CodeAnalysis.TreeAndVersion>:Start<Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13> (Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13&) at AsyncMethodBuilder.cs:471
    frame #7: Mono`Microsoft.CodeAnalysis.DocumentState:IncrementallyParseTreeAsync (Roslyn.Utilities.ValueSource`1<Microsoft.CodeAnalysis.TreeAndVersion>,Roslyn.Utilities.ValueSource`1<Microsoft.CodeAnalysis.TextAndVersion>,System.Threading.CancellationToken) at TextFileUtility.cs:280
    frame #8: 0x000000014b79f5ca Mono`Microsoft.CodeAnalysis.DocumentState/<>c__DisplayClass12_0:<CreateLazyIncrementallyParsedTree>b__0 (System.Threading.CancellationToken) + 42
    frame #9: 0x000000014801f2c8 Mono`Roslyn.Utilities.AsyncLazy`1<T_REF>:StartAsynchronousComputation (Roslyn.Utilities.AsyncLazy`1/AsynchronousComputationToStart<T_REF>,Roslyn.Utilities.AsyncLazy`1/Request<T_REF>,System.Threading.CancellationToken) + 344
    frame #10: 0x000000014801b99f Mono`Roslyn.Utilities.AsyncLazy`1<T_REF>:GetValueAsync (System.Threading.CancellationToken) + 1423
    frame #11: 0x000000014b7a189c Mono`Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13:MoveNext () + 1020
    frame #12: Mono`System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<Microsoft.CodeAnalysis.TreeAndVersion>:Start<Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13> (Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13&) at AsyncMethodBuilder.cs:471
    frame #13: Mono`Microsoft.CodeAnalysis.DocumentState:IncrementallyParseTreeAsync (Roslyn.Utilities.ValueSource`1<Microsoft.CodeAnalysis.TreeAndVersion>,Roslyn.Utilities.ValueSource`1<Microsoft.CodeAnalysis.TextAndVersion>,System.Threading.CancellationToken) at TextFileUtility.cs:280
    frame #14: 0x000000014b79f5ca Mono`Microsoft.CodeAnalysis.DocumentState/<>c__DisplayClass12_0:<CreateLazyIncrementallyParsedTree>b__0 (System.Threading.CancellationToken) + 42
    frame #15: 0x000000014801f2c8 Mono`Roslyn.Utilities.AsyncLazy`1<T_REF>:StartAsynchronousComputation (Roslyn.Utilities.AsyncLazy`1/AsynchronousComputationToStart<T_REF>,Roslyn.Utilities.AsyncLazy`1/Request<T_REF>,System.Threading.CancellationToken) + 344
    frame #16: 0x000000014801b99f Mono`Roslyn.Utilities.AsyncLazy`1<T_REF>:GetValueAsync (System.Threading.CancellationToken) + 1423
    frame #17: 0x000000014b7a189c Mono`Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13:MoveNext () + 1020
    frame #18: Mono`System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<Microsoft.CodeAnalysis.TreeAndVersion>:Start<Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13> (Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13&) at AsyncMethodBuilder.cs:471
    frame #19: Mono`Microsoft.CodeAnalysis.DocumentState:IncrementallyParseTreeAsync (Roslyn.Utilities.ValueSource`1<Microsoft.CodeAnalysis.TreeAndVersion>,Roslyn.Utilities.ValueSource`1<Microsoft.CodeAnalysis.TextAndVersion>,System.Threading.CancellationToken) at TextFileUtility.cs:280
    frame #20: 0x000000014b79f5ca Mono`Microsoft.CodeAnalysis.DocumentState/<>c__DisplayClass12_0:<CreateLazyIncrementallyParsedTree>b__0 (System.Threading.CancellationToken) + 42
    frame #21: 0x000000014801f2c8 Mono`Roslyn.Utilities.AsyncLazy`1<T_REF>:StartAsynchronousComputation (Roslyn.Utilities.AsyncLazy`1/AsynchronousComputationToStart<T_REF>,Roslyn.Utilities.AsyncLazy`1/Request<T_REF>,System.Threading.CancellationToken) + 344
    frame #22: 0x000000014801b99f Mono`Roslyn.Utilities.AsyncLazy`1<T_REF>:GetValueAsync (System.Threading.CancellationToken) + 1423
    frame #23: 0x000000014b7a189c Mono`Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13:MoveNext () + 1020
    frame #24: Mono`System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<Microsoft.CodeAnalysis.TreeAndVersion>:Start<Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13> (Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13&) at AsyncMethodBuilder.cs:471
    frame #25: Mono`Microsoft.CodeAnalysis.DocumentState:IncrementallyParseTreeAsync (Roslyn.Utilities.ValueSource`1<Microsoft.CodeAnalysis.TreeAndVersion>,Roslyn.Utilities.ValueSource`1<Microsoft.CodeAnalysis.TextAndVersion>,System.Threading.CancellationToken) at TextFileUtility.cs:280
    frame #26: 0x000000014b79f5ca Mono`Microsoft.CodeAnalysis.DocumentState/<>c__DisplayClass12_0:<CreateLazyIncrementallyParsedTree>b__0 (System.Threading.CancellationToken) + 42
    frame #27: 0x000000014801f2c8 Mono`Roslyn.Utilities.AsyncLazy`1<T_REF>:StartAsynchronousComputation (Roslyn.Utilities.AsyncLazy`1/AsynchronousComputationToStart<T_REF>,Roslyn.Utilities.AsyncLazy`1/Request<T_REF>,System.Threading.CancellationToken) + 344
    frame #28: 0x000000014801b99f Mono`Roslyn.Utilities.AsyncLazy`1<T_REF>:GetValueAsync (System.Threading.CancellationToken) + 1423
    frame #29: 0x000000014b7a189c Mono`Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13:MoveNext () + 1020
    frame #30: Mono`System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<Microsoft.CodeAnalysis.TreeAndVersion>:Start<Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13> (Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13&) at AsyncMethodBuilder.cs:471
    frame #31: Mono`Microsoft.CodeAnalysis.DocumentState:IncrementallyParseTreeAsync (Roslyn.Utilities.ValueSource`1<Microsoft.CodeAnalysis.TreeAndVersion>,Roslyn.Utilities.ValueSource`1<Microsoft.CodeAnalysis.TextAndVersion>,System.Threading.CancellationToken) at TextFileUtility.cs:280
    frame #32: 0x000000014b79f5ca Mono`Microsoft.CodeAnalysis.DocumentState/<>c__DisplayClass12_0:<CreateLazyIncrementallyParsedTree>b__0 (System.Threading.CancellationToken) + 42
    frame #33: 0x000000014801f2c8 Mono`Roslyn.Utilities.AsyncLazy`1<T_REF>:StartAsynchronousComputation (Roslyn.Utilities.AsyncLazy`1/AsynchronousComputationToStart<T_REF>,Roslyn.Utilities.AsyncLazy`1/Request<T_REF>,System.Threading.CancellationToken) + 344
    frame #34: 0x000000014801b99f Mono`Roslyn.Utilities.AsyncLazy`1<T_REF>:GetValueAsync (System.Threading.CancellationToken) + 1423
    frame #35: 0x000000014b7a189c Mono`Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13:MoveNext () + 1020
    frame #36: Mono`System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<Microsoft.CodeAnalysis.TreeAndVersion>:Start<Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13> (Microsoft.CodeAnalysis.DocumentState/<IncrementallyParseTreeAsync>d__13&) at AsyncMethodBuilder.cs:471
    frame #37: Mono`Microsoft.CodeAnalysis.DocumentState:IncrementallyParseTreeAsync (Roslyn.Utilities.ValueSource`1<Microsoft.CodeAnalysis.TreeAndVersion>,Roslyn.Utilities.ValueSource`1<Microsoft.CodeAnalysis.TextAndVersion>,System.Threading.CancellationToken) at TextFileUtility.cs:280
    frame #38: 0x000000014b79f5ca Mono`Microsoft.CodeAnalysis.DocumentState/<>c__DisplayClass12_0:<CreateLazyIncrementallyParsedTree>b__0 (System.Threading.CancellationToken) + 42
    frame #39: 0x000000014801f2c8 Mono`Roslyn.Utilities.AsyncLazy`1<T_REF>:StartAsynchronousComputation (Roslyn.Utilities.AsyncLazy`1/AsynchronousComputationToStart<T_REF>,Roslyn.Utilities.AsyncLazy`1/Request<T_REF>,System.Threading.CancellationToken) + 344
(lldb) 
```

**Comment by KirillOsenkov** [comment]
I'm not sure this is a Roslyn bug but since the stack ends in AsyncLazy I can't be sure of anything. Let's use this issue to discuss and we can close or move it out if it's not Roslyn.

**Comment by KirillOsenkov** [comment]
@kumpera says it's a stack overflow. Any ideas how can we get into this state? @jasonmalinowski

**Comment by jasonmalinowski** [comment]
Are the AsyncLazys on the stack all the same one or different ones? If they are different, this could be because we ended up with some really long chain of incremental parses to get the tree that overflowed. I believe there is logic to break the chain to prevent us spending more time incremental parsing than a straight parse, but I'm not sure we ever thought about the stack overflow considerations of the design either.

**Comment by KirillOsenkov** [comment]
Yes this is typing at the end of a very long file (string.cs).

**Comment by Pilchie** [comment]
@jasonmalinowski Do you suspect this is an `AsyncLazy/Workspace` issue, or a compiler issue with the incremental parser?

**Comment by jasonmalinowski** [comment]
@Pilchie No idea.

**Comment by Pilchie** [comment]
@KirillOsenkov Any chance you can repro on Windows and get us a crashdump?

**Comment by KirillOsenkov** [comment]
Unfortunately I'm unable to repro...

**Comment by kumpera** [comment]
Maybe it only happens on OSX tha has a shorter default stack size.

**Comment by Pilchie** [comment]
@jaredpar - I suspect your team is more set up to debug issues on Mono than mine.  Any chance you could take a first look here and see if this is the parser proper or the workspace?

**Comment by mattwar** [comment]
This has happened before in a typing-like scenario (lots of small changes) where nothing requests the tree for a long period.  Each change builds up a chain of pending incremental parses.  Eventually, when the request does come, the chain is too long and there is a stack overflow.  This generally doesn't happen in VS typing, since the background parser is always requesting the tree.

**Comment by KirillOsenkov** [comment]
Excellent, looks like this is our scenario (in XS we don't request the tree a lot yet)

**Comment by mattwar** [comment]
Maybe someone with a working roslyn build environment can bring my PR forward and submit it?

**Comment by tmeschter** [comment]
@jasonmalinowski It sounds like we have a fix in hand for this, yes?

**Comment by jasonmalinowski** [comment]
@tmeschter If by "fix in hand" you mean "a fix written by somebody no longer on the team that nobody ever reviewed and who knows if it still compiles", sure.

**Comment by tmeschter** [comment]
@jasonmalinowski Let me rephrase, then: do you think we will be able to fix this for 15.3?

**Comment by mattwar** [comment]
The PR was reviewed at the time... and denied due to difference in opinion on the nature of the problem and whether it should be fixed at all.

I don't expect anyone to just take what I did an pull it into the code base without a review of the problem and determination if my PR is the desired fixed.

But I do have candy.

**Comment by MattGertz** [comment]
Mondo escrow triage: will not meet the escrow bar unless there's evidence that it's being hit widespread, so moving to 15.5.  Tag me if we have data that says otherwise.

**Comment by dpoeschl** [comment]
Moving out because there's been no activity on the [Xamarin bug](https://bugzilla.xamarin.com/show_bug.cgi?id=55597), which was resolved a couple months ago.

**Comment by CyrusNajmabadi** [comment]
Closing out as not repro.  We haven't heard more cases of this.  Note: expectation for any host for the roslny workspace is that they be doing parsing in a timely fashion so that the chain of edits for a file doesn't grow too large.
